run.c

```diff
@@ -186,20 +235,67 @@ void softmax(float* x, int size) {
     }
 }
 
-void matmul(float* xout, float* x, float* w, int n, int d) {
+void matmul(float* xout, int8_t* xq, float* xs, int8_t* wq, float* ws, int n, int d) {
     // W (d,n) @ x (n,) -> xout (d,)
     // by far the most amount of time is spent inside this little function
+    // inputs to this function are both quantized
+
     int i;
     #pragma omp parallel for private(i)
     for (i = 0; i < d; i++) {
+
         float val = 0.0f;
-        for (int j = 0; j < n; j++) {
-            val += w[i * n + j] * x[j];
+        int32_t ival = 0;
+        int in = i * n;
+
+        // do the matmul in groups of GS
+        int j;
+        for (j = 0; j <= n - GS; j += GS) {
+            for (int k = 0; k < GS; k++) {
+                ival += ((int32_t) xq[j + k]) * ((int32_t) wq[in + j + k]);
```

Comment:
Maybe this is would be faster, maybe due to better SIMD
```suggestion
                ival += ((int16_t) xq[j + k]) * ((int16_t) wq[in + j + k]);
```

Reply:
on my linuxbox this does nothing

Reply:
Whether speed improves or not depends, as always. It's logically more correct IMHO. It seems that gcc figures out that mine is actually equivalent to yours and already optimizes "my" way with `vpmullw`:
https://godbolt.org/z/TbdoxT7TT

Maybe `VPDPBUSD` would be faster
https://en.wikichip.org/wiki/x86/avx512_4vnniw).

Reply:
Hmm, it may be possible to extract `VPDPBUSD` from a compiler following either way
https://godbolt.org/z/ccahMeTn5

using the "official" idiom
https://www.intel.com/content/www/us/en/developer/articles/technical/vectorization-llvm-gcc-cpus-gpus.html#gs.411p4x